plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,305 @@
+-- the other generators.
+runGenTmCustom :: Int
+               -> (Maybe (Type TyName DefaultUni ())
+                  -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ()))
+               -> GenTm a
+               -> Gen a
+runGenTmCustom f cg g = do
+  sized $ \ n -> do
+    let env = GenEnv
+          { geSize               = n
+          , geDatas              = Map.empty
+          , geTypes              = Map.empty
+          , geTerms              = Map.empty
+          , geUnboundUsedTyNames = Set.empty
+          , geEscaping           = YesEscape
+          , geCustomGen          = cg
+          , geCustomFreq         = f
+          , geDebug              = False
+          }
+    flip runReader env <$> runGenT g
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+noEscape :: GenTm a -> GenTm a
+noEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
+onSize :: (Int -> Int) -> GenTm a -> GenTm a
+onSize f = local $ \ env -> env { geSize = f (geSize env) }
+
+-- | Default to the first generator if the size is zero (or negative),
+-- use the second generator otherwise.
+ifSizeZero :: GenTm a -> GenTm a -> GenTm a
+ifSizeZero ifZ nonZ = do
+  n <- asks geSize
+  if n <= 0 then ifZ else nonZ
+
+-- | Locally set the size in a generator
+withSize :: Int -> GenTm a -> GenTm a
+withSize = onSize . const
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments.
+sizeSplit_ :: Int -> Int -> GenTm a -> GenTm b -> GenTm (a, b)
+sizeSplit_ a b ga gb = sizeSplit a b ga (const gb)
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments and use the result of the first generator
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> GenTm (a, b)
+sizeSplit a b ga gb = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  (,) x <$> withSize nb (gb x)
+
+-- * Names
+
+-- | Get all uniques we have generated and are used in the current context.
+getUniques :: GenTm (Set Unique)
+getUniques = do
+  GenEnv{geDatas = dts, geTypes = tys, geTerms = tms, geUnboundUsedTyNames = used} <- ask
+  return $ Set.mapMonotonic (_nameUnique . unTyName) (Map.keysSet dts <> Map.keysSet tys <> used) <>
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
+   names are not immediately put into any state when generated. There is *no guarantee*
+   that in this situation:
+   ```
+   do nms <- genFreshNames ss
+      nms' <- genFreshNames ss
+   ```
+   the names in `nms` and `nms'` don't overlap.
```

Comment:
The problem with using a state monad is that you still need to wrap your code with "register binding" because you need to unregister the binding when it's out of scope again. So a reader is more natural for bindings.

Reply:
You're absolutely right, I don't know what I was thinking. I'll resolve all my related comments.